boot/seal.go

```diff
@@ -25,65 +25,269 @@ import (
+	}
+	if len(snaps) != 1 {
+		return bc, fmt.Errorf("cannot obtain recovery kernel snap")
+	}
+	seedKernel := snaps[0]
+
+	var kernelRev string
+	if seedKernel.SideInfo.Revision.Store() {
+		kernelRev = seedKernel.SideInfo.Revision.String()
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildRecoveryAssetChain(rbl, modeenv)
+	if err != nil {
+		return bc, err
+	}
+
+	return bootChain{
+		BrandID:        model.BrandID(),
+		Model:          model.Model(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+		AssetChain:     assetChain,
+		Kernel:         seedKernel.Path,
+		KernelRevision: kernelRev,
+		KernelCmdlines: []string{cmdline},
+		model:          model,
+		kernelBootFile: bootloader.NewBootFile(seedKernel.Path, kbf.Path, kbf.Role),
+	}, nil
+}
+
+func buildRunModeBootChains(rbl, bl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv, cmdline string) ([]bootChain, error) {
+	// get asset chains
+	assetChain, kbf, err := buildRunModeAssetChain(rbl, bl, modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	// get run mode kernels
+	runModeKernels, err := runModeKernelsFromModeenv(modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	chains := make([]bootChain, 0, 2)
+	for _, k := range runModeKernels {
+		info, err := snap.ParsePlaceInfoFromSnapFileName(k)
+		if err != nil {
+			return nil, err
+		}
+		var kernelRev string
+		if info.SnapRevision().Store() {
+			kernelRev = info.SnapRevision().String()
+		}
+		chains = append(chains, bootChain{
+			BrandID:        model.BrandID(),
+			Model:          model.Model(),
+			Grade:          string(model.Grade()),
+			ModelSignKeyID: model.SignKeyID(),
+			AssetChain:     assetChain,
+			Kernel:         k,
+			KernelRevision: kernelRev,
+			KernelCmdlines: []string{cmdline},
+			model:          model,
+			kernelBootFile: bootloader.NewBootFile(k, kbf.Path, kbf.Role),
+		})
+	}
+
+	return chains, nil
+}
+
+func buildRecoveryAssetChain(rbl bootloader.Bootloader, modeenv *Modeenv) (assets []bootAsset, kernel bootloader.BootFile, err error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, kernel, fmt.Errorf("bootloader doesn't support trusted assets")
+	}
+
+	recoveryBootChain, err := tbl.RecoveryBootChain("")
+	if err != nil {
+		return nil, kernel, err
+	}
+
+	// the last entry is the kernel
+	numAssets := len(recoveryBootChain) - 1
+	assets = make([]bootAsset, numAssets)
+
+	for i := 0; i < numAssets; i++ {
+		name := filepath.Base(recoveryBootChain[i].Path)
+		hashes, ok := modeenv.CurrentTrustedRecoveryBootAssets[name]
+		if !ok {
+			return nil, kernel, fmt.Errorf("cannot find asset %s in modeenv", name)
+		}
+		assets[i] = bootAsset{
+			Role:   recoveryBootChain[i].Role,
+			Name:   name,
+			Hashes: hashes,
+		}
+	}
+
+	return assets, recoveryBootChain[numAssets], nil
+}
+
+func buildRunModeAssetChain(rbl, bl bootloader.Bootloader, modeenv *Modeenv) (assets []bootAsset, kernel bootloader.BootFile, err error) {
```

Comment:
This and buildRecoveryAssetChain can take  bootloader's RecoveryBootChain (or BootChain) and be one function.

Reply:
This is done now.